ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,895 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+	aggr->match_extnl_hash = NULL;
+
+	XFREE(MTYPE_OSPF_EXTERNAL_RT_AGGR, aggr);
+}
+
+static void ospf_external_aggr_add(struct ospf *ospf,
+				   struct ospf_external_aggr_rt *aggr)
+{
+	struct route_node *rn;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Adding Aggregate route to Aggr table (%s/%d)",
+			   __func__, inet_ntoa(aggr->p.prefix),
+			   aggr->p.prefixlen);
+	rn = route_node_get(ospf->rt_aggr_tbl, (struct prefix *)&aggr->p);
+	if (rn->info)
+		route_unlock_node(rn);
+	else
+		rn->info = aggr;
+}
+
+static void ospf_external_aggr_delete(struct ospf *ospf, struct route_node *rn)
+{
+	struct ospf_external_aggr_rt *aggr = rn->info;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Deleting Aggregate route (%s/%d)", __func__,
+			   inet_ntoa(aggr->p.prefix), aggr->p.prefixlen);
+
+	/* Sent a Max age LSA if it is already originated. */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Flushing Aggregate route (%s/%d)",
+				   __func__, inet_ntoa(aggr->p.prefix),
+				   aggr->p.prefixlen);
+		ospf_external_lsa_flush(ospf, 0, &aggr->p, 0);
+	}
+
+	rn->info = NULL;
+	route_unlock_node(rn);
+	route_unlock_node(rn);
+}
+
+struct ospf_external_aggr_rt *
+ospf_extrenal_aggregator_lookup(struct ospf *ospf, struct prefix_ipv4 *p)
+{
+	struct route_node *rn;
+
+	rn = route_node_lookup(ospf->rt_aggr_tbl, (struct prefix *)p);
+	if (rn) {
+		route_unlock_node(rn);
+		return rn->info;
+	}
+	return NULL;
+}
+
+struct ospf_external_aggr_rt *ospf_external_aggr_match(struct ospf *ospf,
+						       struct prefix_ipv4 *p)
+{
+	struct route_node *node;
+
+	node = route_node_match(ospf->rt_aggr_tbl, (struct prefix *)p);
+	if (node) {
+
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			if (node->info) {
+				struct ospf_external_aggr_rt *ag = node->info;
+
+				zlog_debug(
+					"%s: Matching aggregator found.prefix:%s/%d Aggregator %s/%d\n",
+					__func__, inet_ntoa(p->prefix),
+					p->prefixlen, inet_ntoa(ag->p.prefix),
+					ag->p.prefixlen);
+			}
+
+		route_unlock_node(node);
+		return node->info;
+	}
+	return NULL;
+}
+
+void ospf_unlink_ei_from_aggr(struct ospf *ospf,
+			      struct ospf_external_aggr_rt *aggr,
+			      struct external_info *ei)
+{
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug(
+			"%s: Unlinking extrenal route(%s/%d) from aggregator(%s/%d), external route count:%ld",
+			__func__, inet_ntoa(ei->p.prefix), ei->p.prefixlen,
+			inet_ntoa(aggr->p.prefix), aggr->p.prefixlen,
+			OSPF_EXTERNAL_RT_COUNT(aggr));
+	hash_release(aggr->match_extnl_hash, ei);
+	ei->aggr_route = NULL;
+
+	/* Flush the aggreagte route if matching
+	 * external route count becomes zero.
+	 */
+	if (!OSPF_EXTERNAL_RT_COUNT(aggr)
+	    && CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Flushing the aggreagte route (%s/%d)",
+				   __func__, inet_ntoa(aggr->p.prefix),
```

Comment:
use `%pI4` to print ipv4 addresses rather than `inet_ntoa`

Reply:
done